### ISSUE TITLE
ofEventListener: new method isSubscribed()

### ISSUE THREAD

**Issue by bakercp** [openframeworks/openFrameworks#5646]
It would be useful to know if an `ofEventListener` is currently subscribed to something.

Currently we can explicitly `unsubscribe()`, but there it would be nice to check something like:

```c++

class ofEventListener{
public:
	ofEventListener(){}
	ofEventListener(std::unique_ptr<of::priv::AbstractEventToken> && token)
	:token(std::move(token)){}
	void unsubscribe(){
		token.reset();
	}

        bool isSubscribed() const
        {
               return token != nullptr;
        }
private:
	std::unique_ptr<of::priv::AbstractEventToken> token;
};

```

**Comment by roymacdonald** [comment]
Hi, I'd been writing an ofBook chapter about events, and noticed some wording issues.
There is `ofAddListener(...)`, `ofRemoveListener(...),` `ofRegisterMouseEvents()`, `ofUnregisterMouseEvents()`, and now in the `ofEventListener` we have `unsubscribe()`.
In order to reduce the use of new words, why don't we deprecate `unsubscribe()` and add `unregister()`. And for what's proposed here by @bakercp  (which I think it is a must) lets use `isRegistered()`
What do you say?
ping @arturoc

**Comment by bakercp** [comment]
I'm all for deprecating and condensing.  I just want a way to tell if it is subscribed/registered.

In thinking about the right words to describe the function though, I feel like "subscribe" is more descriptive (to a new user) than "register".  

Usually one "registers" an event if there is some sense of a global event queue and then after the event is registered, listeners can "subscribe" to that event.

Since we don't really have a "registration" process for our events (they are usually just members of classes, etc), it seems that subscribe better describes what we're doing when we add an event listener.

**Comment by roymacdonald** [comment]
ha, my english falls short.
I get you are saying, I was proposing this just to avoid more words.
Following your logic, shouldn't `ofRegisterMouseEvents()` be `ofSubscribeMouseEvents()`?
Not really sure now.... :$

**Comment by bakercp** [comment]
Yes, it should probably be ... but usually it's "subscribe to" so maybe `ofSubscribeToMouseEvents()` but that feels a bit weird.  I don't know.  Need more input from others I think :)

**Comment by roymacdonald** [comment]
Yeah I thought of that too, and I agree, its a bit weird.
Lets wait for input. 
or maybe this is being too picky. ;)

best!

**Comment by roymacdonald** [comment]
Now I need this! :)
I had to add it to my own repo.
It is absolutely needed in order to avoid subscribing twice.
all the best!

**Comment by arturoc** [comment]
i'm not super opposed to have this but i've never used it and i can imagine in some multi threaded environments it wouldn't even be accurate (ofEvent is thread safe so having a thread safe api is at least desirable). 

Also when you overwrite a listener the old one is unregistered right away because the old ofEventListener is destroyed so subscribing twice should never happen in any case what do you mean that is absolutely needed to avoid subscribing twice?

**Comment by roymacdonald** [comment]
Hi arturo,
the threading issue makes sense.
I wasn't super clear. I can guess that if a listener is already subscribed it will make no harm to register once again using the same parameters.
My problem is that, in my specific case, I would need to  store a bool just for tracking if the listener is subscribed or not, so having this method avoids this, making the code cleaner, easier to read and safer. I need to use it this way because in some cases, along with registering the listener I need to reset/setup some other stuff. Resetting these again once already register brings problems.

It kinda makes sense to me to have this method and if it is not harmful then I see no problems having it.
What would involve making it thread safe? Just adding a scoped lock ?

best

**Comment by arturoc** [comment]
mmh it seems then the problem is in keeping track of the mutable state of your own class which among other things involves registering a listener but the registration of the listener itself is not the problem

there's no way to make this thread safe cause right after you've asked if it's registered it could have been registered from a different thread, it's an edge case but i really don't see the need for a method like this when you can just register the event again and again without causing any problem.

**Comment by roymacdonald** [comment]
Well it is not a real problem for me to keep track of the mutable state, rather it "feels better" to have the whole thing encapsulated in the listener. 
It kinda makes sense to me that a class (ofEventListener) that changes its state (subscribe and unsubscribe) has a way of querying its state, for whichever purpose.

> there's no way to make this thread safe cause right after you've asked if it's registered it could have been registered from a different thread, it's an edge case

I don't see why it would make it thread unsafe. (Here I lack of deeper c++ knowledge). I'm just curious to know why.
cheers
